File: m4d/ClientApp/src/models/KeywordQuery.ts
const regex = /(?<field>Artist|Title|Albums):\((?<search>[^)]*)\)/g;

export class KeywordQuery {
  public static fromParts(parts: Map<string, string>): KeywordQuery {
    let simple = true;
    const segments = [];
    const everywhere = parts.get("Everywhere") ?? "";

    parts.forEach((value, key) => {
      if (value && key !== "Everywhere") {
        segments.push(`${key}:(${value})`);
        simple = false;
      }
    });
    if (everywhere) {
      segments.push(everywhere);
    }
    return new KeywordQuery(simple ? everywhere : "`" + segments.join(" "));
  }

  public constructor(s?: string) {
    this.data = s ?? "";
  }

  public get isLucene(): boolean {
    return this.data.startsWith("`");
  }

  public get search(): string {
    return this.isLucene ? this.data.substring(1) : this.data;
  }

  public get query(): string {
    return this.data;
  }

  public update(part: string, value: string): KeywordQuery {
    const parts = this.fields;
    if (value) {
      parts.set(part, value);
    } else {
      parts.delete(part);
    }
    return KeywordQuery.fromParts(parts);
  }

  public get description(): string {
    if (this.data === "") {
      return "";
    }

    if (!this.isLucene) {
      return `containing the text "${this.data}"`;
    }

    const fields = this.fields;
    const all = fields.get("Everywhere");
    fields.delete("Everywhere");

    if (all && fields.size === 0) {
      return `containing the text "${all}"`;
    }

    let result = "where";
    let first = true;
    if (all) {
      result = `containing the text "${all}" anywhere`;
      first = false;
    }

    fields.forEach((value, key) => {
      if (!first) {
        result += " and";
      }
      result += ` ${key.toLowerCase()} contains "${value}"`;
      first = false;
    });

    return result;
  }

  public getField(field: string): string {
    return this.fields.get(field) ?? "";
  }

  private get fields(): Map<string, string> {
    const search = this.search;
    const matches = search.matchAll(regex);
    const result = new Map<string, string>();
    for (const match of matches) {
      result.set(match.groups?.field ?? "", match.groups?.search ?? "");
    }
    const all = search.replaceAll(regex, "").trim();
    if (all) {
      result.set("Everywhere", all);
    }
    return result;
  }

  private data: string;
}

File: m4dModels/KeywordQuery.cs
using System.Collections.Generic;
using System.Text.RegularExpressions;

namespace m4dModels
{

    public class KeywordQuery(string s = "")
    {
        private readonly string data = s ?? "";

        public static KeywordQuery FromParts(Dictionary<string, string> parts)
        {
            var simple = true;
            var segments = new List<string>();
            var everywhere = parts.TryGetValue("Everywhere", out var value) ? value : "";

            foreach (var part in parts)
            {
                if (!string.IsNullOrEmpty(part.Value) && part.Key != "Everywhere")
                {
                    segments.Add($"{part.Key}:({part.Value})");
                    simple = false;
                }
            }

            if (!string.IsNullOrEmpty(everywhere))
            {
                segments.Add(everywhere);
            }

            return new KeywordQuery(simple ? everywhere : "`" + string.Join(" ", segments));
        }

        public bool IsLucene => data.StartsWith('`');

        public string Keywords
        {
            get { return IsLucene ? data[1..] : data; }
        }

        public string Query
        {
            get { return data; }
        }

        public KeywordQuery Update(string part, string value)
        {
            var parts = Fields;
            if (!string.IsNullOrEmpty(value))
            {
                parts[part] = value;
            }
            else
            {
                parts.Remove(part);
            }
            return FromParts(parts);
        }

        public string Description
        {
            get
            {
                if (string.IsNullOrEmpty(data))
                {
                    return "";
                }

                if (!IsLucene)
                {
                    return $"containing the text \"{data}\"";
                }

                var fields = Fields;
                var all = fields.TryGetValue("Everywhere", out var value) ? value : "";
                fields.Remove("Everywhere");

                if (!string.IsNullOrEmpty(all) && fields.Count == 0)
                {
                    return $"containing the text \"{all}\"";
                }

                var result = "where";
                var first = true;
                if (!string.IsNullOrEmpty(all))
                {
                    result = $"containing the text \"{all}\" anywhere";
                    first = false;
                }

                foreach (var @field in fields)
                {
                    if (!first)
                    {
                        result += " and";
                    }
                    result += $" {field.Key.ToLower()} contains \"{field.Value}\"";
                    first = false;
                }

                return result;
            }
        }

        public string ShortDescription
        {
            get
            {
                if (string.IsNullOrEmpty(data))
                {
                    return "";
                }

                if (!IsLucene)
                {
                    return $"\"{data}\"";
                }

                var fields = Fields;
                var all = fields.TryGetValue("Everywhere", out var value) ? value : "";
                fields.Remove("Everywhere");

                if (!string.IsNullOrEmpty(all) && fields.Count == 0)
                {
                    return $"\"{all}\"";
                }

                var result = "";
                var first = true;
                if (!string.IsNullOrEmpty(all))
                {
                    result = $"\"{all}\" anywhere";
                    first = false;
                }

                foreach (var @field in fields)
                {
                    if (!first)
                    {
                        result += " and ";
                    }
                    result += $"\"{field.Value}\" in {field.Key.ToLower()}";
                    first = false;
                }

                return result;
            }
        }

        public string GetField(string field)
        {
            return Fields.TryGetValue(field, out var value) ? value : "";
        }

        private Dictionary<string, string> Fields
        {
            get
            {
                var search = Keywords;
                var matches = _regex.Matches(search);
                var result = new Dictionary<string, string>();
                foreach (Match match in matches)
                {
                    result[match.Groups["field"].Value] = match.Groups["search"].Value;
                }
                var all = _regex.Replace(search, "").Trim(); // replace "regex" with your actual regex pattern
                if (!string.IsNullOrEmpty(all))
                {
                    result["Everywhere"] = all;
                }
                return result;
            }
        }

        private static readonly Regex _regex = new(@"(?<field>Artist|Title|Albums):\((?<search>[^)]*)\)", RegexOptions.Compiled);
    }
}
